crates/zome_types/src/lib.rs

```diff
@@ -73,3 +75,121 @@ pub trait CallbackResult {
     /// answer like a fail, and we don't want to over-optimise wasm calls and miss a clear failure
     fn is_definitive(&self) -> bool;
 }
+
+#[macro_export]
+/// Serialization for fixed arrays is generally not available in a way that can be derived.
+/// Being able to wrap fixed size arrays is important e.g. for crypto safety etc. so this is a
+/// simple way to implement serialization so that we can send these types between the host/guest.
+macro_rules! fixed_array_serialization {
+    ($t:ty, $len:expr) => {
+        impl serde::ser::Serialize for $t {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                serializer.serialize_bytes(&self.0)
+            }
+        }
+
+        impl<'de> serde::de::Deserialize<'de> for $t {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: serde::de::Deserializer<'de>,
+            {
+                use serde::de::Error;
+                let bytes: &[u8] = serde::de::Deserialize::deserialize(deserializer)?;
+                if bytes.len() != $len {
+                    let exp_msg = format!("expected {} bytes got: {} bytes", $len, bytes.len());
+                    return Err(D::Error::invalid_value(
+                        serde::de::Unexpected::Bytes(bytes),
+                        &exp_msg.as_str(),
+                    ));
+                }
+                let mut inner: [u8; $len] = [0; $len];
+                inner.clone_from_slice(bytes);
+                Ok(Self(inner))
+            }
+        }
+    };
+}
+
+#[macro_export]
+/// Cryptographic secrets are fiddly at the best of times.
+///
+/// In wasm it is somewhat impossible to have true secrets because wasm memory is not secure.
+///
+///  - The host can always read wasm memory so any vulnerability in the host compromises the guest.
+///  - The host/rust generally doesn't guarantee to immediately wipe/zero out freed memory, either
+///    when a zome call is running or after a wasm instance is thrown away.
+///
+/// Most of the time we should just try to minimise the interaction between wasm and secret data.
+///
+/// For example, lair keeps all our private keys internal and we can only send it signing requests
+/// associated with public keys.
+///
+/// In other contexts it is more difficult, such as when generating secrets from raw cryptographic
+/// random bytes and sending them to peers directly.
```

Comment:
From an infosec perspective, you should just never do this. This is the point of asynchronous crypto - individuals generate keypairs and keep their private keys private, only publishing public keys.

I worry about publishing an API that encourages bad practices.